src/main.rs

```diff
@@ -486,17 +486,173 @@ fn shell_exec(vm: &VirtualMachine, source: &str, scope: Scope) -> ShellExecResul
+
+            // last: the last word, could be empty if it ends with a dot
+            // parents: the words before the dot
+
+            let mut current = self.scope.load_global(self.vm, first)?;
+
+            for attr in parents {
+                current = self.vm.get_attribute(current.clone(), attr.as_str()).ok()?;
+            }
+
+            (
+                Box::new(str_iter(
+                    self.vm.call_method(&current, "__dir__", vec![]).ok()?,
+                )?) as StrIter,
+                last.as_str(),
+            )
+        } else {
+            // we need to get a variable based off of globals/builtins
+
+            let globals = str_iter(
+                self.vm
+                    .call_method(self.scope.globals.as_object(), "keys", vec![])
+                    .ok()?,
+            )?;
+            let iter = if first.as_str().is_empty() {
+                // only show globals that don't start with a  '_'
+                Box::new(globals.filter(|r| {
+                    r.as_ref()
+                        .ok()
+                        .map_or(true, |s| !s.as_str().starts_with('_'))
+                })) as StrIter
+            } else {
+                // show globals and builtins
+                Box::new(
+                    globals.chain(str_iter(
+                        self.vm
+                            .call_method(&self.vm.builtins, "__dir__", vec![])
+                            .ok()?,
+                    )?),
+                ) as StrIter
+            };
+            (iter, first.as_str())
+        };
+        let completions = iter
+            .filter(|res| {
+                res.as_ref()
+                    .ok()
+                    .map_or(true, |s| s.as_str().starts_with(prefix))
+            })
+            .collect::<Result<Vec<_>, _>>()
+            .ok()?;
+        let no_underscore = completions
+            .iter()
+            .cloned()
+            .filter(|s| !prefix.starts_with('_') && !s.as_str().starts_with('_'))
+            .collect::<Vec<_>>();
+        let mut completions = if no_underscore.is_empty() {
+            completions
+        } else {
+            no_underscore
+        };
+        completions.sort_by(|a, b| std::cmp::Ord::cmp(a.as_str(), b.as_str()));
+        Some((
+            startpos,
+            completions
+                .into_iter()
+                .map(|s| s.as_str().to_owned())
+                .collect(),
+        ))
+    }
+}
+
+impl rustyline::completion::Completer for ShellHelper<'_> {
+    type Candidate = String;
+
+    fn complete(
+        &self,
+        line: &str,
+        pos: usize,
+        _ctx: &rustyline::Context,
+    ) -> rustyline::Result<(usize, Vec<String>)> {
+        if pos != line.len() {
+            return Ok((0, vec![]));
+        }
+        Ok(self.complete_opt(line).unwrap_or((0, vec![])))
+    }
+}
+
+impl rustyline::hint::Hinter for ShellHelper<'_> {}
+impl rustyline::highlight::Highlighter for ShellHelper<'_> {}
```

Comment:
Oh cool! Does this thing support syntax highlights?

Reply:
Yep, it does! It'd probably be pretty easy to plug a Python [`syntect`](https://github.com/trishume/syntect) definition in for this.